### DISCUSSION TITLE
Building for or installing on 8088 and NEC V20?

### DISCUSSION THREAD

**Discussion by hilga007** [EtchedPixels/FUZIX#1072] [General]
Long story short - PC/XT Clone with 8088 and 8087. Plan to upgrade to a V20 and install a new crystal. RAM max at 640K. Want to use Soundblaster 8-bit ISA cards, IDE, TV Tuner, better graphics, etc, and FUSIX seems like it would offer the most "familiar" UNIXy experience.

Any luck installing on Intel 8088 platforms? Which compiler is needed?

Ideal scenario is to install onto the NEC V20 with 640K RAM and a 1.5MB RAM ISA upgrade card + Ethernet + ISA --> USB card and the upgraded Trident video card from various homebrew hardware projects. 

I'm really excited to dive in.

**Comment by EtchedPixels** [comment]
I've done some initial work on 8088 but that is all. I would take a look at ELKS, as that's back actively alive and a well developed 8086 Linux subset including networking.

https://github.com/ghaerr/elks

That's one reason I've not really done much with x86.

**Reply by hilga007** [reply]
> I've done some initial work on 8088 but that is all. I would take a look at ELKS, as that's back actively alive and a well developed 8086 Linux subset including networking.
> 
> 
> 
> https://github.com/ghaerr/elks
> 
> 
> 
> That's one reason I've not really done much with x86.
> 
> 

True, though they seem to hard-target 16-bit. A couple of Discord servers shuffled me back over to you to consider the 8008, 8080, 8085 and 8088. 

Lemme go confirm with The ELKS homies!!!

**Reply by ghaerr** [reply]
Hello @hilga007,

Over at ELKS we're targeting all variants of the real mode 16-bit x86 (8088, 8086, NEC V20, 8018X, 80386 real mode etc) segmented architecture processors, but not 8-bit CPUs like the 8080, 8085 or Z80.

Thank you!

**Reply by hilga007** [reply]
Thank you @ghaerr  and @EtchedPixels for your super fast replies. You are an epic mini-*NIX TEAM! WOW! Thanks for considering my midday (for me) musings...

**Reference:** it looks like the 8008, 8080, 8085, 8088, Z80, eZ80 all share one thing in common: 16-bit address width (or 12), and 8-bit data bus. 

So it seems like: 

- **FUZIX**: 8008 8080, 8085 _(and variants)_
- **FUZIX AND ELKS:** 8088 _(FUZIX has experience with 8-bit data busses that could result in an ELKS/FUZIX OS for that model and its variants (only suggesting because you seem like a Q&A duo today!)_
- **ELKS**: 8086 and beyond

(I've been doodling this list for far too long to try to wrap my head around the scope of my ask, as well as the net societal benefit) 

**8008**: 12-bit address width, 8-bit data bus. Lineup includes: 

- Intel 8008, C8008, D8008, G8008, C8008-1 D8008-1 
- Siemens SAB8008-1C
- Microsystems International MF8008R, MF8008-1R
- DDR U808D (East German Clone)

**8080**: 16-bit address width, 8-bit data bus. Lineup includes: 

- Intel 8080, 8080A-1, 8080A-2
- NEC μCOM-8 (μPD753) - 8080 compatible 
- NEC μCOM-80 (μPD8080A) - 8080 compatible, extra custom and Z80 instructions
- NEC μCOM-80F (μPD8080AF) - 8080 compatible, extra inst. removed
- AMD AM9080ADC/C8080A AKA 9080 (Intel 8080 compatible) 
- National Semiconductor INS8080AN, NTE NTE8080A
- Signetics MP8080AI
- Texas Instruments TMS8080JL
- Toshiba TMP9080AP

**8085**: 16-bit address width, 8-bit data bus. Lineup Includes: 

- Intel 8085, 8085A, P8085AH, 8085-2, 8085A-2, P8085AH-2, P8085AH-1, ID8085, M8085A
- Intel 80C85
- NEC μCOM-85 (μPD8085)- 8085 compatible 
- **WHAT!?** _Russian Still Manufactured 80C85 Clone: ИМ1821ВМ85А_
- Mitsubishi M5L8085 - 8085 compatible 
- OKI M80C85
- Siemens SAB8085
- Toshiba TMP8085

**8088**: 16-bit address width, 8-bit data bus. Lineup Includes: 

- Intel 8088 
- NEC V20 
- NEC μCOM-88 (μPD8088) 
- AMD QM8088D1
- Fujitsu MBL8088-C, MBL8088 (ceramic)
- MHS P8088
- OKI M80C88A-10 (QFP)
- Siemens SAB8088-12-P

**Reply by ghaerr** [reply]
You've got a good list of 16-bit address/8 bit data bus CPUs, but it's not just how the hardware addresses memory that counts, when it comes to system software. 

While the 8088 grabs data by the byte, its internal ISA (instruction set architecture) is 16-bits wide, with 4 segment registers that actually allow addressing 20 bits (1Mb) of memory. In the 8-bit ISA world, there are also differences between processor instruction sets.

For instance, ELKS will run on an 8088 or 8086 with no system software changes, as the ISA is identical. This sets the 8088 and 8086 apart from the 8080, 8085 and Z80, as the ISA is quite different, although similar enough to allow programmers of that era to more quickly understand how to port their software to the newer CPU. The 8088 -> 8086 hardware transition similarly effectively allowed hardware designers to upgrade their designs without having to toss their 8-bit bus designs immediately.

The 8080, 8085 and Z80 have similar but not identical ISAs to each other, and that also allowed programmers to port their software quickly. However, the 8080 and 8088/8086 are quite different.

**Reply by EtchedPixels** [reply]
8086/88 take you up a size so to speak. You've easily got 128K of memory usable per program whereas on the older processors you tend to have to stuff each application into 32-48K. That alone means the way things are done and what works best changes. Elks has access to a lot of more featured versions of tools than Fuzix as a result.

**Reply by hilga007** [reply]
Thank you both again so much for educating me! Even though I dislike not knowing something I want to know about, I am a very slow learner and your kindness and helpfulness really go a long way here. 

 So that means ELKS, and always ELKS is the way to go for 8086/8088! 
 
 And that FUZIX is always the way to go for 8008/8080/8085/Z80/eZ80/Z180 :) 
 
 MUCH appreciated!!! :)
